From 81d392acb2a9adcb7a86d2abca537e526de17b06 Mon Sep 17 00:00:00 2001
From: Christopher Albert <albert@tugraz.at>
Date: Fri, 14 Nov 2025 11:46:16 +0100
Subject: [PATCH] fortran: align spec-stmt message with verify_st_order

* parse.cc (parse_executable): Reword diagnostic to say "%s
  statement ... cannot follow executable statement" mirroring other
  ordering errors.
* testsuite/gfortran.dg/spec_statement_in_exec.f90: Expect updated
  wording.

PR fortran/32365

Signed-off-by: Christopher Albert <albert@tugraz.at>
---
 gcc/fortran/parse.cc                          |  4 ++--
 .../gfortran.dg/spec_statement_in_exec.f90    | 24 +++++++++----------
 2 files changed, 14 insertions(+), 14 deletions(-)

diff --git a/gcc/fortran/parse.cc b/gcc/fortran/parse.cc
index 96382512884..3784b2a00cf 100644
--- a/gcc/fortran/parse.cc
+++ b/gcc/fortran/parse.cc
@@ -7132,10 +7132,10 @@ loop:
 	  accept_statement (st);
 	  goto done;
 
-	/* Specification statements cannot appear after executable statements.  */
+	/* Specification statements cannot follow executable statements.  */
 	case_decl:
 	case_omp_decl:
-	  gfc_error ("%s statement at %C after the first executable statement",
+	  gfc_error ("%s statement at %C cannot follow executable statement",
 		     gfc_ascii_statement (st));
 	  reject_statement ();
 	  st = next_statement ();
diff --git a/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90 b/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90
index 268304bc049..397c57c94ff 100644
--- a/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90
+++ b/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90
@@ -11,18 +11,18 @@ subroutine test_spec_in_exec
   i = 1
 
   ! Test key specification statement types
-  integer :: j                     ! { dg-error "data declaration statement cannot appear after an executable statement" }
-  real :: x                       ! { dg-error "data declaration statement cannot appear after an executable statement" }
-  complex :: z                    ! { dg-error "data declaration statement cannot appear after an executable statement" }
-  logical :: flag                  ! { dg-error "data declaration statement cannot appear after an executable statement" }
-  character(len=20) :: name       ! { dg-error "data declaration statement cannot appear after an executable statement" }
-  double precision :: d           ! { dg-error "data declaration statement cannot appear after an executable statement" }
-  common /myblock/ i              ! { dg-error "COMMON statement cannot appear after an executable statement" }
-  equivalence (i, i)              ! { dg-error "EQUIVALENCE statement cannot appear after an executable statement" }
-  namelist /nml/ i                ! { dg-error "NAMELIST statement cannot appear after an executable statement" }
-  data i / 2 /                    ! { dg-error "DATA statement cannot appear after an executable statement" }
-!$omp threadprivate(i)             ! { dg-error "THREADPRIVATE statement cannot appear after an executable statement" }
-!$omp declare target (i)           ! { dg-error "DECLARE TARGET statement cannot appear after an executable statement" }
+  integer :: j                     ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  real :: x                       ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  complex :: z                    ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  logical :: flag                  ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  character(len=20) :: name       ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  double precision :: d           ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  common /myblock/ i              ! { dg-error "COMMON statement at \\(1\\) cannot follow executable statement" }
+  equivalence (i, i)              ! { dg-error "EQUIVALENCE statement at \\(1\\) cannot follow executable statement" }
+  namelist /nml/ i                ! { dg-error "NAMELIST statement at \\(1\\) cannot follow executable statement" }
+  data i / 2 /                    ! { dg-error "DATA statement at \\(1\\) cannot follow executable statement" }
+!$omp threadprivate(i)             ! { dg-error "THREADPRIVATE statement at \\(1\\) cannot follow executable statement" }
+!$omp declare target (i)           ! { dg-error "DECLARE TARGET statement at \\(1\\) cannot follow executable statement" }
 
 end subroutine test_spec_in_exec
 
-- 
2.51.2

